lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1433,14 +1438,24 @@ class DifferentiableActivityInfo {
   void setVaried(SILValue value, unsigned independentVariableIndex);
   void setVariedAcrossArrayInitialization(SILValue value,
                                           unsigned independentVariableIndex);
+  /// Marks the given value as varied and propagates variedness to users.
+  void setVariedAndPropagateToUsers(SILValue value,
+                                    unsigned independentVariableIndex);
+  /// Marks the given instruction as having propagated variedness for the given
+  /// independent variable index.
+  void setDidPropagateVaried(SILInstruction *inst,
+                             unsigned independentVariableIndex);
+  /// Propagates variedness for the given instruction, from operands to results.
+  void propagateVaried(SILInstruction *inst, unsigned independentVariableIndex);
+  /// Marks the given value as varied and recursively propagates variedness
+  /// inwards (to operands) through projections. Skips `@noDerivative` struct
```

Comment:
> [...] And what are non-differentiable address projections?

Non-differentiable address projections include:
1. `struct_element_addr` to a `@noDerivative` field.
2. `tuple_element_addr` to an element with a non-differentiable type, like `Int`.

Example of case 2:
```swift
@_silgen_name("nondifferentiable_tuple_element_addr")
@differentiable
func bar(_ x: Float) -> Float {
  var tuple: (Int, Int, Float, Float) = (1, 1, x, x)
  tuple.0 = 1 // non-differentiable `tuple_element_addr`
  tuple.0 = 1 // non-differentiable `tuple_element_addr`
  tuple.3 = x
  return tuple.3
}
```

Such `tuple_element_addr` projections are currently not marked as active (not marked as varied, to be precise):
```
[AD] Activity info for nondifferentiable_tuple_element_addr at (source=0 parameters=(0))
bb0:
[ACTIVE] %0 = argument of bb0 : $Float                     // users: %37, %18, %17, %1
[ACTIVE]   %2 = alloc_stack $(Int, Int, Float, Float), var, name "tuple" // users: %43, %3, %4, %5, %6, %23, %31, %35, %39
[USEFUL]   %3 = tuple_element_addr %2 : $*(Int, Int, Float, Float), 0 // user: %11
[USEFUL]   %4 = tuple_element_addr %2 : $*(Int, Int, Float, Float), 1 // user: %16
[ACTIVE]   %5 = tuple_element_addr %2 : $*(Int, Int, Float, Float), 2 // user: %17
[ACTIVE]   %6 = tuple_element_addr %2 : $*(Int, Int, Float, Float), 3 // user: %18
...
```

> What's the justification for skipping `@noDerivative` struct field projections and non-differentiable address projections? [...] Aren't we supposed to produce a diagnostic when variedness flows through something non-differentiable?

Good question. Here's how the differentiation transform currently works:
* Activity analysis runs before derivative function generation, determining which values in the original function need a derivative.
* Then, [`LinearMapInfo::shouldDifferentiateInstruction`](https://github.com/apple/swift/blob/56548e0ff250ac6b69de5fed023ddf672f761e3a/lib/SILOptimizer/Mandatory/Differentiation.cpp#L1594) uses activity info to determine which instructions to visit during [differential](https://github.com/apple/swift/blob/56548e0ff250ac6b69de5fed023ddf672f761e3a/lib/SILOptimizer/Mandatory/Differentiation.cpp#L4530)/[pullback](https://github.com/apple/swift/blob/56548e0ff250ac6b69de5fed023ddf672f761e3a/lib/SILOptimizer/Mandatory/Differentiation.cpp#L6625) generation.
* If `LinearMapInfo::shouldDifferentiateInstruction` returns true for an instruction, then the instruction is visited during differential/pullback generation. Non-differentiability diagnostics are produced in instruction visitors, e.g. [active `apply` `inout` arguments](https://github.com/apple/swift/blob/56548e0ff250ac6b69de5fed023ddf672f761e3a/lib/SILOptimizer/Mandatory/Differentiation.cpp#L5421). (Non-differentiability diagnostics are also produced in places other than instruction visitors.)

Rather than "producing a diagnostic when variedness flows through something non-differentiable", we currently "avoid marking non-differentiable projections as active, so they're not even visited by differential/pullback generation".

Semantically, this makes sense because non-differentiable projections indeed do not need a derivative, so diagnostics should not be emitted for [writes to non-differentiable projections](https://github.com/apple/swift/pull/27472/files#diff-da8b8c9bee81d7af57b10d63a2641b4bR47).

"Not marking non-differentiable projections as active" also avoids extra downstream work for the compiler and enables nice simplifications, like [making `PullbackEmitter::getAdjointBuffer` no longer need to check for `@noDerivative` or emit diagnostics](https://github.com/apple/swift/pull/27472).

---

I think not marking non-differentiable projections as active is the right direction, and changes like https://github.com/apple/swift/pull/27472 move us towards the desired non-differentiability diagnostics. But the implementation has room for improvement! TF-947 tracks improvements for variedness propagation, additional ideas are welcome!

Reply:
Right, non-differentiable projects (esp. the explicit ones marked by the library with `@noDerivative`) should never be active.

Reply:
@marcrasi pointed out that "skipping `@noDerivative` struct projections during variedness propagation" may not be consistent with accessors, which are used in cross-module cases.

I added an item to TF-947 to track checking whether activity analysis is consistent for `struct_extract`/`struct_element_addr` projections and accessor functions to `@noDerivative` properties.